lblod/ember-rdfa-editor
# Add ability to manage document-external triples

### Overview


Add ability to add external (= loose) triples to a resource node

External triples are data not connected to an in-document node. They can be used
to define knowledge about URIs which are not mentioned in the document. 

It's expected this will be rather rare, but crucial for some embedded usecases.

External triples are serialized inside a new container inside the already
exising rdfa-container, annotated with the `data-external-triple-container` 
attribute, which is why I don't consider this a breaking change. 

Also adds a new transactionMonad for manipulating these:
`transformExternalTriples`, and 2 shorthand wrappers: `addExternalTriple` and
`setExternalTriples`.

It also expands the rdfa-editor sidebar tools with an external triple section.

##### connected issues and PRs:



https://binnenland.atlassian.net/browse/GN-5170
[plugin pr](https://github.com/lblod/ember-rdfa-editor-lblod-plugins/pull/510)
[embeddable pr](https://github.com/lblod/frontend-embeddable-notule-editor/pull/269)

### Setup



### How to test/reproduce

There's a few automated tests that hopefully document a bit how it works. But should be fairly straightforward, you add triples and they stick around when you reload the doc, and otherwise don't get in the way of anything

I've done my best to keep the commits reasonably atomic, but everything is called metaTriples before the rename commit, keep that in mind if you're browsing the history.
### Challenges/uncertainties

- It relies too much on a sort of quirk of the rdfa-processing logic for my liking, that whole thing needs a more robust implementation, but that'd likely involve writing our own parser.

- has somewhat weird but also kind of logical behavior when you mention the same resource in 2 different nodes: it collects all the knowledge about each subject onto all nodes that talk about it

kind of nice in a way, but also kind of not, cause if you edit in one location, you now have conflicting data.
But I don't have time to make document-wide editing logic. It's starting to feel a lot like a document-agnostic way of adding/editing information is in the cards though, which would be really cool.


- I know external-triple-form duplicates a lot of boilerplate from outgoing-triple-form, but I couldn't find a nice extraction point in a reasonable amount of time, as it ultimately comes down to building a form library

### Checks PR readiness
